net/src/packet.rs

```diff
@@ -0,0 +1,251 @@
+        let mut this = Packet {
+            eth: eth.clone(),
+            net: None,
+            transport: None,
+            vlan: ArrayVec::default(),
+            net_ext: ArrayVec::default(),
+        };
+        let mut prior = Header::Eth(eth);
+        loop {
+            let header = prior.step(&mut cursor);
+            match prior {
+                Header::Eth(eth) => this.eth = eth,
+                Header::Ipv4(ip) => this.net = Some(Net::Ipv4(ip)),
+                Header::Ipv6(ip) => this.net = Some(Net::Ipv6(ip)),
+                Header::Tcp(tcp) => this.transport = Some(Transport::Tcp(tcp)),
+                Header::Udp(udp) => this.transport = Some(Transport::Udp(udp)),
+                Header::Icmp4(icmp4) => this.transport = Some(Transport::Icmp4(icmp4)),
+                Header::Icmp6(icmp6) => this.transport = Some(Transport::Icmp6(icmp6)),
+                Header::Vlan(vlan) => {
+                    if this.vlan.len() < MAX_VLANS {
+                        this.vlan.push(vlan);
+                    } else {
+                        break;
+                    }
+                }
+                Header::IpAuth(auth) => {
+                    if this.net_ext.len() < MAX_NET_EXTENSIONS {
+                        this.net_ext.push(NetExt::IpAuth(auth));
+                    } else {
+                        break;
+                    }
+                }
+                Header::IpV6Ext(ext) => {
+                    if this.net_ext.len() < MAX_NET_EXTENSIONS {
+                        this.net_ext.push(NetExt::Ipv6Ext(ext));
+                    } else {
+                        break;
+                    }
+                }
+            }
+            match header {
+                None => {
+                    break;
+                }
+                Some(next) => {
+                    prior = next;
+                }
+            }
+        }
+        #[allow(unsafe_code)] // Non zero checked by parse impl
+        let consumed = unsafe { NonZero::new_unchecked(cursor.inner.len() - cursor.remaining) };
+        Ok((this, consumed))
+    }
+}
+
+impl DeParse for Packet {
+    type Error = ();
+
+    fn size(&self) -> NonZero<usize> {
+        // TODO(blocking): Deal with ip{v4,v6} extensions
+        let eth = self.eth.size().get();
+        let vlan = self.vlan.iter().map(|v| v.size().get()).sum::<usize>();
+        let net = match self.net {
+            None => {
+                debug_assert!(self.transport.is_none());
+                0
+            }
+            Some(ref n) => n.size().get(),
+        };
+        let transport = match self.transport {
+            None => 0,
+            Some(ref t) => t.size().get(),
+        };
+        NonZero::new(eth + vlan + net + transport).unwrap_or_else(|| unreachable!())
+    }
+
+    fn write(&self, buf: &mut [u8]) -> Result<NonZero<usize>, DeParseError<Self::Error>> {
+        // TODO(blocking): Deal with ip{v4,v6} extensions
```

Comment:
Why is there no abstraction here like there is for `step` on parse?  Please create an issue to fix this and address that issue right after the PR that has a main that parses a packet.

Reply:
This is a good point and is arguably a design miss.  I will think about good ways to fix this